### ISSUE TITLE
Feature Request: Archive old file versions

### ISSUE THREAD

**Issue by lhoenig** [Emily3403/isisdl#15]
Hey, an idea for a new feature: it happens from time to time that courses update already uploaded files. When running `isisdl`, the new file is downloaded and the old version is overwritten as far as I can see. Would it be possible for `isisdl` to pick up on the fact that an older version of the file has already been downloaded and move it something like `<basename-wo-ext>-<datetime>.<ext>` instead of overwriting it?

This behavior could then also be made configurable.

**Comment by Emily3403** [comment]
This is a very good idea and one that is definitely possible to implement as far as I can tell*. I'll try to roll it out as part of the 1.4 update.

*This feature will not work for all types of files. If a file is uploaded directly to ISIS (`==MediaType.document`), then the size is visible and it can be determined if a new version is available. Unfortunately, this will not work for external links due to the fact that no size information is available (anymore). For the most part (~90%) it will work perfectly though.

**Comment by lhoenig** [comment]
Awesome!! thanks so much.

**Comment by lhoenig** [comment]
Regarding the difficulties: I see often the upload date on ISIS, is this information not visible in the Moodle API?
For example:
<img width="493" alt="Screenshot 2022-10-28 at 15 43 40" src="https://user-images.githubusercontent.com/4353030/198620395-a8f536a6-75ce-48bd-8bfe-d1c85fd8a7c2.png">
The filesize is not a super reliable indicator of whether the file has changed or not. I understand that the upload date information may not be available for all the ways files can be uploaded to ISIS. But then you could still download the new file to a tmpdir and check the shasum to determine if the file has changed. Same for the external files. In conjunction with this, it may be useful to save the `isisdl` download-date in the database for renaming the file, or simply use the creation date on supported filesystems.

**Comment by Emily3403** [comment]
This information is available through the API in particular `timecreated` and `timemodified`. These _could_ be used as indicators if a file has changed. Currently, I only store a single attribute `time`, which indicates the last time the file was modified. 

This attribute, however, is not mandatory anymore in the new 1.4 version and, interestingly enough, it could now be used as a indicator if a file has changed because in the 1.3 version the fallback was the current date. And for some files that were extern it would always fill in the current datetime. 

The file size is actually a pretty good indicator if two files are different. From my testing there were only a few courses (from *all* courses of ISIS) that had collisions with file sizes. In fact this mechanism is the steering force between the ability to check for conflicts and resolve them. Thus I am pretty confident that it is reliable _enough_. Of course, it is not perfect, but in practice it should be good enough.

But both approaches could be used in a combined approach: First check if either attribute diverges. If it is the size, then the file must be different and it is replaced and archived. In case that only the time, but not the size is divergent, the file is downloaded in a temporary directory and verified against the existing checksum to determine if it is the same. If it is not the same it is replaced and archived and if it is the same, this fact could be noted in the database to prevent future replacing (with an exponential backoff strategy of some sort).

**Comment by lhoenig** [comment]
Thanks for all the further infos! The combined approach sounds like a good idea, but I'd keep it simple at first.
As a sidenote, my opinion is that the filesize alone should never be used to conclude that two files are identical (if you don't know anything about the file contents) - there's just so much that can go wrong, even if we don't see it so much in the current content of ISIS courses. Checking the size first for improved speed is a nice idea however.

**Comment by lhoenig** [comment]
Before you write it yourself - See https://docs.python.org/3/library/filecmp.html#module-filecmp :) That sounds pretty much like the function you describe.

**Comment by Emily3403** [comment]
Interesting that you bring up this module because I had a look at it myself when I decided I needed checksums of a file as a function. The problem I was solving originally is verifying file integrity. This is very similar to checking if two files are the same as you are comparing the old to the current version. From there I coded an algorithm that can calculate a checksum of the file in either $\mathcal{O}(\log(n))$ or in $\mathcal{O}(n)$. It spits out a `.hexdigest()` of the file and that is stored in the database along with the other meta-information.

As a by-product of this functionality it can be dected if two files are the same. This part should (hopefully) be pretty straightforward. The only problem is deciding when to check the two files …

**Comment by lhoenig** [comment]
Yeah, I also just yesterday learned of the existence of `filecmp` :)
If you have such an algorithm already, then of course it makes sense to use it. The function `filecmp.cmp` just implements the logic of first checking the filesize, and if it differs then performing a buffered read / compare of the files until EOF or a difference is found. It also has a cache. But as the cache size is limited to 100, `isisdl` will frequently exceed it anyways.

**Comment by Emily3403** [comment]
After implementing a bunch of stuff in the 2.0 branch, I can say that this feature is still planned and is actively being developed. It does make the DB Model quite a bit more complicated as now a single URL might correspond to multiple files, however I feel like it is worth it.
